src/runtime_src/xdp/profile/plugin/aie_trace/aie_trace_plugin.cpp

```diff
@@ -481,86 +475,101 @@ namespace xdp {
-        if (ret != XAIE_OK) break;
-        ret = perfCounter->reserve();
-        if (ret != XAIE_OK) break;
 
-        // Set reset event based on counter number
-        XAie_Events counterEvent;
-        perfCounter->getCounterEvent(mod, counterEvent);
-        int idx = static_cast<int>(counterEvent) - static_cast<int>(XAIE_EVENT_PERF_CNT_0_MEM);
-        perfCounter->changeThreshold(memoryCounterEventValues.at(i));
+        for (int i=0; i < memoryCounterStartEvents.size(); ++i) {
+          auto perfCounter = memory.perfCounter();
+          auto ret = perfCounter->initialize(mod, memoryCounterStartEvents.at(i),
+                                             mod, memoryCounterEndEvents.at(i));
+          if (ret != XAIE_OK) 
+            break;
+          ret = perfCounter->reserve();
+          if (ret != XAIE_OK) 
+            break;
 
-        perfCounter->changeRstEvent(mod, counterEvent);
-        memoryEvents.push_back(counterEvent);
+          // Set reset event based on counter number
+          XAie_Events counterEvent;
+          perfCounter->getCounterEvent(mod, counterEvent);
+          int idx = static_cast<int>(counterEvent) - static_cast<int>(XAIE_EVENT_PERF_CNT_0_MEM);
+          perfCounter->changeThreshold(memoryCounterEventValues.at(i));
 
-        ret = perfCounter->start();
-        if (ret != XAIE_OK) break;
+          perfCounter->changeRstEvent(mod, counterEvent);
+          memoryEvents.push_back(counterEvent);
 
-        mMemoryCounters.push_back(perfCounter);
-        numMemoryCounters++;
+          ret = perfCounter->start();
+          if (ret != XAIE_OK) 
+            break;
 
-        // Update config file
-        uint8_t phyEvent = 0;
-        auto& cfg = cfgTile->memory_trace_config.pc[idx];
-        XAie_EventLogicalToPhysicalConv(aieDevInst, loc, mod, memoryCounterStartEvents[i], &phyEvent);
-        cfg.start_event = phyEvent;
-        XAie_EventLogicalToPhysicalConv(aieDevInst, loc, mod, memoryCounterEndEvents[i], &phyEvent);
-        cfg.stop_event = phyEvent;
-        XAie_EventLogicalToPhysicalConv(aieDevInst, loc, mod, counterEvent, &phyEvent);
-        cfg.reset_event = phyEvent;
-        cfg.event_value = memoryCounterEventValues[i];
-        cfgTile->memory_trace_config.packet_type=1;
+          mMemoryCounters.push_back(perfCounter);
+          numMemoryCounters++;
+
+          // Update config file
+          uint8_t phyEvent = 0;
+          auto& cfg = cfgTile->memory_trace_config.pc[idx];
+          XAie_EventLogicalToPhysicalConv(aieDevInst, loc, mod, memoryCounterStartEvents[i], &phyEvent);
+          cfg.start_event = phyEvent;
+          XAie_EventLogicalToPhysicalConv(aieDevInst, loc, mod, memoryCounterEndEvents[i], &phyEvent);
+          cfg.stop_event = phyEvent;
+          XAie_EventLogicalToPhysicalConv(aieDevInst, loc, mod, counterEvent, &phyEvent);
+          cfg.reset_event = phyEvent;
+          cfg.event_value = memoryCounterEventValues[i];
+        }
```

Comment:
This is quite a function!
While working on it, it might be good to split these huge functions into pieces.
When browsing the code it looks like there are already similar repeated code.

Reply:
To be updated in future PR